File: Graph/c++/03_팀결성.cpp
// 이것이 취업을 위한 코딩테스트다 298p
// 1

#include <iostream>
#include <vector>

using namespace std;

int n, m;
int parent[100001];

int findParent(int x)
{
	if (x == parent[x]) return x;
	return parent[x] = findParent(parent[x]);
}

void unionParent(int a, int b)
{
	a = findParent(a);
	b = findParent(b);
	if (a < b) parent[b] = a;
	else parent[a] = b;
}

int main()
{
	cin >> n >> m;
	
	for (int i = 1; i <= n; i++)
		parent[i] = i;

	for (int i = 0; i < m; i++)
	{
		int a, b, c;
		cin >> a >> b >> c;
		if (a == 0)
			unionParent(b, c);
		else if (a == 1)
		{
			if (findParent(b) == findParent(c))
				cout << "YES" << '\n';
			else
				cout << "NO" << '\n';
		}
	}

	return 0;
}

File: Graph/python/03_팀결성.py
# 이것이 취업을 위한 코딩테스트다 298p
# 2

N, M = map(int, input().split())
parent = [0] * (N + 1)
for i in range(1, N + 1):
    parent[i] = i

def find_parent(x):
    if parent[x] != x:
        parent[x] = find_parent(parent[x])
    return parent[x]

def union_parent(a, b):
    a = find_parent(a)
    b = find_parent(b)
    if a < b:
        parent[b] = a
    else:
        parent[a] = b

for _ in range(M):
    oper, a, b = map(int, input().split())
    if oper == 0:
        union_parent(a, b)
    else:
        if find_parent(a) == find_parent(b):
            print("YES")
        else:
            print("NO")